ELEC-E5550 - Statistical Natural Language Processing
# SET 1: Text Preprocessing

## Released: 19.01.2021 at 14:30
## Deadline: 01.02.2021 at 23:59

# Overview
Consider this assignment an introduction to statistics of different language units (letters, pairs of letters, words). We will explore the frequency distribution of these different language units and then discuss what this knowledge might give. Moreover, we'll talk about how to handle raw text, how to separate it into different units and how those units and operations are called.

# Table of contents

* [Task 1: Letter and Letter pair Frequency analysis](#task_1)
    * [Step 1.1: Prepare the text](#subtask_1_1)
    * [Step 1.2: Get letter frequencies](#subtask_1_2)
    * [Step 1.3: Letter frequency analysis](#subtask_1_3)
    * [Step 1.4: Count all possible two-letter strings](#subtask_1_4)
    * [Step 1.5: Get letter pair counts](#subtask_1_5)
    * [Step 1.6: Letter pair frequency analysis](#subtask_1_6)
* [Task 2: Word Tokenization](#task_2)
    * [Step 2.1: Tokenize by whitespaces](#subtask_2_1)
    * [Step 2.2: Tokenize with regular expressions](#subtask_2_2)
    * [Step 2.3: Use Treebank tokenizer](#subtask_2_3)
* [Task 3: ](#task_3)
    * [Step 3.1: Analyse word frequencies](#subtask_3_1)
    * [Step 3.2: Remove stop words](#subtask_3_2)
    * [Step 3.3: How much is left without stop words?](#subtask_3_3) 
* [Checklist before submission](#checklist)

## TASK 1 <a class="anchor" id="task_1"></a>
## Letter and Letter pair Frequency analysis

The data used in this assignment is "The Gold-Bug" by Edgar Allan Poe. It is actually a story about the importance of letter frequencies. The narrator in the story was able to decipher a message leading to a hidden treasure by applying frequency analysis. The cipher used in the story is a substitute cipher where each letter is replaced by a different letter or number.

Knowing the frequency of letters in a language is important not only for solving ciphers, but it also has practical applications like data compression. For example, Morse code uses the shortest symbols for the most frequent letters. 

In this task you'll need to discover for yourself the frequency distribution of single letters and of letter pairs in English.

## 1.1  <a class="anchor" id="subtask_1_1"></a>
### Prepare the text (1 point)

First of all, we need to load the text into the Jupyter Notebook and prepare it for further analysis. We will focus only on letters now and we don't care if they are capitalized or lowercased. 

Create a function that reads the data located in `/coursedata/01-text-processing/the_gold-bug.txt`, leaves only lowercased alphabetic characters in it, and writes them into a list as separate letters.

HINT1: [string methods might come in handy](https://www.w3schools.com/python/python_ref_string.asp)

HINT2: you can employ Python's open() and read() functions

HINT3: Alphabetic characters are characters defined as “Letter” in the Unicode character database

In [14]:
def read(file_name):
    """This function creates a list of lowercase alphabetic characters from a .txt file 
    
    Parameters
    ----------
    file_name : str
        a path to the text file
    
    Returns
    -------
    lowercased_letters : list of strings
        text as a list of lowercase letters
    """
    lowercased_letters=list()
    with open(file_name, 'r') as fr:
        for line in fr:
            for char in line:
                if char.isalpha():# char!='\n' and char!=' ' an:
                    lowercased_letters.append(char.lower())
        #[ch for ch in open('test.txt').read() if ch != '\n' if ch != ' ']
        #string=fr.read()#.replace('\n', '')
        #print(string)
    #lowercased_letters=string.lower()
    # YOUR CODE HERE
    
    
    return lowercased_letters

bug_letters = read('/coursedata/01-text-processing/the_gold-bug.txt')

In [15]:
from nose.tools import assert_equal

# checks if your function returns a list
assert_equal(type(read('/coursedata/01-text-processing/the_gold-bug.txt')), list)

# checks if your function returns a list of strings
assert_equal(type(read('/coursedata/01-text-processing/the_gold-bug.txt')[0]), str)

# checks if your list is of the right length
assert_equal(len(read('/coursedata/01-text-processing/the_gold-bug.txt')), 58269)

# checks if the strings in your list are alpabetic characters
assert_equal(all([letter.isalpha() for letter in read('/coursedata/01-text-processing/the_gold-bug.txt')]), True)

# checks if the strings in your list are lowercased
assert_equal(all([letter.islower() for letter in read('/coursedata/01-text-processing/the_gold-bug.txt')]), True)

# checks if your list has the first 10 members right 
assert_equal(read('/coursedata/01-text-processing/the_gold-bug.txt')[:10], 
             ['t', 'h', 'e', 'g', 'o', 'l', 'd', 'b', 'u', 'g'])


## 1.2 <a class="anchor" id="subtask_1_2"></a>
### Get letter frequencies (3 points)

Now we can count how many times each letter occurred in the story, and then turn these counts into Maximum Likelihood probability estimates of seeing each letter. 

To do so, write a function that takes in a list of letters and returns a dictionary with their frequencies relative to the size of the whole text: $ freq_x = \frac{n_x}{N} $, where $n_x$ is the number of times a letter $x$ was seen, and $N$ is the total number of letters in the text. These relative frequencies are probability estimates for the letters in our text.

HINT: you might find **nltk.FreqDist** or **collections.Counter** useful

In [16]:
import collections

def get_freq(letters):
    """This function computes MLE probabilities of letters
    
    Given a list of letters, this function should return a probability dictionary,
    where letters are keys and their MLE probabilities are values

    Parameters
    ----------
    letters : list of strings
        text as a list of only lowercase letters 
    
    Returns
    -------
    letter_freq_dict : dictionary-like object
        a frequency dictionary of letters where values are relative frequencies (MLE probabilities)
    """
    letter_freq_dict=collections.Counter()
    N=len(letters)
    for letter in letters:
        letter_freq_dict[letter]+=1/N
    
    return letter_freq_dict

bug_probability_dict = get_freq(bug_letters)

In [17]:
from numpy.testing import assert_almost_equal

# checks if the alphabet length is 26
assert_equal(len(get_freq(bug_letters)), 26)

# checks if probability of all the letters equals one
assert_almost_equal(sum(get_freq(bug_letters).values()), 1., 3)

# checks if the algorithm is doing what it is supposed to be doing on a dummy example
assert_equal(get_freq(['b','b','b','a','a','c']), {"b":3/6,"a":2/6,"c":1/6})

# checks if the probability of 'e' is correct
assert_almost_equal(get_freq(bug_letters)['e'], 0.13125332509567694, 3)


## 1.3 <a class="anchor" id="subtask_1_3"></a>
### Letter frequency analysis (1 point)
The counts of letters in a text differ quite much. That means the probabilities of seeing each letter are also different.

Using your frequency dictionary, answer the following questions:

1. what is the probability of the most frequent letter? (0.3 points)
2. what is the least probable letter? (0.3 points)
3. what is the order of English letters according to their probability? Answer with a sorted string of letters starting with the most frequent one. (0.4 points)

Type your answers in the cell below. You can create an additional cell to do the calculations if needed. 

In [18]:
# YOUR CODE HERE
sorted_list=sorted(bug_probability_dict.items(), key=lambda pair: pair[1], reverse=True)

# put your answer to question 1 as a float number into the variable below
# For example:
# most_frequent_letter_prob = 0.123
most_frequent_letter_prob = sorted_list[0][1] ## FILL IN THE ANSWER

# put your answer to question 2 as a string into the variable below
# For example:
# least_probable_letter = 'a'
least_probable_letter = sorted_list[0][0] ## FILL IN THE ANSWER

# put your answer to question 3 as a string into the variable below
# For example:
# sorted_letters = 'abcdefghigklmnopqrstuvwxyz'
sorted_letters=""
for char in sorted_list:
    sorted_letters+=char[0]
print(sorted_letters)

etaoinshrdlucmfwpygbvkxjqz


In [19]:
# checks if your answer is a float
assert_equal(type(most_frequent_letter_prob), float)

In [20]:
# checks if your answer is a string
assert_equal(type(least_probable_letter), str)

In [21]:
# checks if you typed in a string of 26 letters
assert_equal(type(sorted_letters), str)
assert_equal(len(sorted_letters), 26)
# checks if the 3rd most frequent letter is 'a'
assert_equal(sorted_letters[2], 'a')

## 1.4 <a class="anchor" id="subtask_1_4"></a>
### Count all possible two-letter strings (1 point)
Some combinations of language units are more likely than other too. You'll see it in a minute.

There are 26 letters in English alphabet. How many possible two-letter strings are there according to combinatorics ('permutations' combinatorically speaking)? For example, if we have an alphabet of 3 letters **a**, **b** and **c**. We can have 9 two-letter strings: **aa**, **bb**, **cc**, **ab**, **ac**, **ba**, **bc**, **ca**, **cb**.

Complete the function below, so we can count all n-length strings for an alphabet of any length.

In [22]:
def number_of_permutatuions(number_of_letters, sequence_len):
    """ Counts the number of possible letter permutations in a string.
    
    This function takes a number of letters in an alphabet and the desired length of a string,
    and outputs the number of all possible letter permutations in a string of this length.
    A string can contain the same letter "sequence_len" times.
    
    Parameters
    ----------
    number_of_letters : int 
        a number of letters in an alphabet
    sequence_len : int
        the length of a string
    
    Returns
    -------
    num_of_permutatuions : int
        the number of all possible strings of a given length with a given alphabet
    """
    
    # YOUR CODE HERE
    num_of_permutations=number_of_letters**(sequence_len)
    return num_of_permutations

In [23]:
assert_equal(number_of_permutatuions(2,2), 4)
assert_equal(number_of_permutatuions(2,1), 2)
assert_equal(number_of_permutatuions(3,2), 9)


## 1.5 <a class="anchor" id="subtask_1_5"></a>
### Get letter pair counts (3 points)

Same as with single letters, some sequences of language units are more probable than others. Not all strings of two letters are actually possible in English. A fact like this can be used in such applications as predictive texting: your phone suggests what might be the next word you need based on previous words you typed (not all word sequences are possible or equally probable).

In the following task, you'll need to count all two-letter strings that appeared in the text.

In [24]:
def count_and_sort_pairs(letters):
    i=0
    pairs=collections.Counter()
    while i<len(letters)-1:
        string=letters[i]+letters[i+1]
        pairs[string]+=1
        i+=1
    pairs_sorted=sorted(pairs.items(), key=lambda pair: pair[1], reverse=True)
    
    """ This function counts letter pairs and sorts them according to their frequency.
    
    This function takes a text represented as a list of lowercase letters
    and converts it into a sorted list of tuples, where the first element
    is a two-letter string, and the second element is the count of this letter pair in the text.
    The first element of a list should be a tuple for the most frequent pair.
    
    Parameters
    ----------
    letters : list of strings
        text as a list of lowercase letters 
    
    Returns
    -------
    pairs_sorted : list of (str, int) tuples
        a list of tuples (letter_pair, count) sorted by the count element 
    """
    
    
    return pairs_sorted

bug_pairs_sorted = count_and_sort_pairs(bug_letters)

In [25]:
# checks if the function returns a list
assert_equal(type(count_and_sort_pairs(bug_letters)), list)

# checks if the function returns a list of tuples
assert_equal(type(count_and_sort_pairs(bug_letters)[0]), tuple)

# checks if the function returns a list of tuples (str, int)
assert_equal((type(count_and_sort_pairs(bug_letters)[0][0]),type(count_and_sort_pairs(bug_letters)[0][1])), (str,int))

# checks that the most frequent pair was seen 1800 times
assert_equal(count_and_sort_pairs(bug_letters)[0][1], 1800)

# checks if a functions works right for the dummy example
assert_equal(count_and_sort_pairs("aaaabbabc")[:2], [("aa",3),("ab",2)])
assert(count_and_sort_pairs("aaaabbabc")[2] in [("ba",1), ("bb", 1), ("bc", 1)])


## 1.6 <a class="anchor" id="subtask_1_6"></a>
### Letter pair frequency analysis (3 points)
Using the sorted list you've created, answer the following questions in the cell below:

1. How many different two-letter combinations have you actually encountered in the data? (0.6 points)
2. What fraction of all theoretically possible two-letter strings is it? (0.6 points)
3. What is the most frequent two-letter string in English? (make a mental note if it is surprising or not) (0.6 points)
4. What is the probability of seeing a pair where both letters are the same? (0.6 points)
5. What is the probability of a pair starting with 'm'? (0.6 points)

Type your answers in the cell below. You can create an additional cell to do the calculations if needed.

In [26]:
# YOUR CODE HERE

# put your answer to question 1 as an int into the variable below
# For example:
# n_pairs = 1000
n_pairs = len(bug_pairs_sorted) ##FILL IN THE ANSWER
print(n_pairs)

# put your answer to question 2 as a float into the variable below
# For example:
frac_of_pairs = n_pairs/(len(sorted_letters)**2) ##FILL IN THE ANSWER
print(frac_of_pairs)
# put your answer to question 3 as a string into the variable below
# For example:
# most_frequent_pair = 'ab'
most_frequent_pair = bug_pairs_sorted[0][0] ##FILL IN THE ANSWER
print(most_frequent_pair)
# put your answer to question 4 as a float into the variable below
# For example:
p_same_letters = 0
for pair in bug_pairs_sorted:
    if pair[0][0]==pair[0][1]:
        p_same_letters+=float(pair[1])
p_same_letters=p_same_letters/len(bug_pai)

print(p_same_letters/(len(bug_letters)-1))
# put your answer to question 5 as a float into the variable below
# For example:
# p_starts_with_m = 0.5
m_letters = 0
for pair in bug_pairs_sorted:
    if pair[0][0]=='m':
        m_letters+=float(pair[1])
p_starts_with_m = m_letters/(len(bug_letters)-1) ##FILL IN THE ANSWER
print(p_starts_with_m)

519
0.7677514792899408
th
0.03459875060067275
0.025725955927781975


In [49]:
# checks if your answer is an int
assert_equal(type(n_pairs) , int)

In [50]:
# checks if your answer is a float
assert_equal(type(frac_of_pairs), float)

In [51]:
# checks if your answer is a string
assert_equal(type(most_frequent_pair),str)

In [52]:
# checks if your answer is a float
assert_equal(type(p_same_letters),float)

In [53]:
# checks if your answer is a float
assert_equal(type(p_starts_with_m),float)

In [29]:
#assert_almost_equal(p_same_letters, 0.03459875060067275, 3)
p_same_letters

2016.0

## TASK 2 <a class="anchor" id="task_2"></a>
## Word Tokenization
In this task, you will create a function that splits the text into more elaborate units than just letters: words


Text data is a part of virtually any NLP application. Sometimes you're lucky, and instead of plain raw text you get nice and clean one, but this is not always the case. Before getting your hands dirty with your actual application, you would most probably need to perform some manipulation to the text. For instance, separate it into words and sentences, remove unwanted symbols. Different tasks require different preprocessing techniques. In this task we'll use some simple ones.

It's not trivial to separate words from a string of text. The first thing that needs to be decided is what to count as a word. Should punctuation and numbers be considered words? Should *frogs* and *frog* be considered the same word? What about *Frog*, *frog* and *FROG*? Before answering those questions, let's make sure we are on the same page and discuss some terminology.

When talking about words, we can mean several different things: lemmas, word types and word tokens.

* **Lemma** - an identifier of a set of lexical forms sharing the same stem (*run* is the lemma for *runs* and *running*), a dictionary form of a word.
* **Word type** - a distinct unit in a text (all the instances of *runs* are counted once).
* **Word token** - every instance of word occurrence (every instance of *runs* counted as a separate word token).

Thus:
* **Tokenization** - a process of separating out word tokens from text
* **Lemmatization** - a process of assigning a group of word forms their lemma, and further separating out these lemmas from text

Generally, English doesn't require lemmatization since it has quite a limited number of word forms. For this reason, we'll leave this task out, for now, and focus on tokenization instead.

Let's create a tokenizer that considers numbers and punctuation as tokens and doesn't separate hyphenated words like *dum-dum*. For that you'll need:
- regular expressions
- string operations

## 2.1 <a class="anchor" id="subtask_2_1"></a>
### Tokenize by whitespaces (1 point)
Let's start off by separating words just by whitespaces and see what happens to our dummy sentence example: 
*It's a dum-dum example, we'll place it here to prove a point. Also look at this number: 300.99.*

HINT: One string method is particularly useful

In [54]:
dummy_example = "It's a dum-dum example, we'll place it here to prove a point. Also, look at this number: 300.99."

def whitespace_tokenize(raw_string):    
    """This function tokenizes strings by whitespaces.
    
    Any whitespace separator should work. 
    For example, this function should be able to tokenize by '\n',
    and two consecutive whitespaces should be regarded as a single separator.
    
    Parameters
    ----------
    raw_string : str
        some text to tokenize
        
    Returns
    -------
    whitespace_tokenized : list of strings
        list of tokens 
    """
    # YOUR CODE HERE
    whitespace_tokenized=raw_string.split()
    
    return whitespace_tokenized

dum_dum_example = whitespace_tokenize(dummy_example)
# see what you've got
print(dum_dum_example)

["It's", 'a', 'dum-dum', 'example,', "we'll", 'place', 'it', 'here', 'to', 'prove', 'a', 'point.', 'Also,', 'look', 'at', 'this', 'number:', '300.99.']


In [55]:
# checks if the first token is correct
assert_equal(dum_dum_example[0], "It's")

# checks if number of tokens is correct
assert_equal(len(dum_dum_example), 18)

# checks if all whitespaces were removed
assert_equal(any([' ' in t for t in dum_dum_example]), False)

# checks if double whitespaces are removed
assert_equal(whitespace_tokenize('  a  a  '), ['a','a'])

# checks if tab ia removed too
assert_equal(whitespace_tokenize('  a \t a  '), ['a','a'])



## 2.2 <a class="anchor" id="subtask_2_2"></a>
### Tokenize with regular expressions (5 points)

As can be seen from the dummy example, it's not enough to just separate words by the whitespaces. This way we get tokens like *'example,'*, *'point.'* and *'number:'*. It's not ideal because we would actually like to have punctuation marks as separate tokens but keep them inside the items like prices and numbers (4.99). Thus, we need something more complex: a regular expression.


If you've never used regular expressions before or you've never used them in Python, you can read about how they work with the re module [here](https://docs.python.org/3/howto/regex.html). 

But to give you a simple example, you can think of a regular expression as a shoe that only fits some of the strings. For example, regular expression 'a' (`re.compile("a")`) only fits a string 'a', but regular expression with a special character '\d' (`re.compile("\d")`) fits any digit '0','1','2',...'9'. Regular expression 'a|\d' (`re.compile("a|\d")`) fits to 'a' string OR to any string with a digit ('|' plays a role of disjunction). You can ask to find all the substrings of a string that match some regular expression pattern with the following piece of code: `re.findall(regular_expression, string_to_look_for_matches)`. More examples are in the cell below.

In [56]:
# example of regular expression usage
import re

regex_digit = re.compile("\d") # any digit
regex_a = re.compile("a") # only letter 'a'
regex_a_digit = re.compile("a\d") # letter 'a' followed by any digit 
regex_a_or_digit = re.compile("a|\d") # letter 'a' OR any digit 
regex_digit_once_or_more = re.compile("\d+") # any digit one or more times

print(re.findall(regex_digit ,'a123hnd7hjaf'))
print(re.findall(regex_a ,'a123hnd7hjaf'))
print(re.findall(regex_a_digit ,'a123hnd7hjaf'))
print(re.findall(regex_a_or_digit ,'a123hnd7hjaf'))
print(re.findall(regex_digit_once_or_more ,'a123hnd7hjaf'))

['1', '2', '3', '7']
['a', 'a']
['a1']
['a', '1', '2', '3', '7', 'a']
['123', '7']


We can use regular expressions to describe what type of substring we want to get out of a string. In addition to separation by whitespaces, we want to keep words or numbers with a hyphen, an apostrophe or a point inside, and we want to split punctuation marks from the end of words. For these means, you'll need to write a regular expression that matches:

- all alphanumeric strings with hyphen, apostrophe or point inside (i.e should be able to find "44.44","a-ha","it's")

**OR**
- any non-whitespace character followed between zero and unlimited times by any alphanumeric character (this part disregards whitespaces and separates punctuation from the end of words)

HINT1: "\S" - non-whitespace character 

HINT2: "\w" - alphanumeric character

HINT3: "[123]" - one of the characters in the brackets

HINT4: "*" - zero or more times

HINT5: more hints in this [cheat sheet](https://www.rexegg.com/regex-quickstart.html)

In [57]:
# YOUR CODE HERE

# put a compiled a regular expression tokenizer into a variabe below
# For example:
# regex_tokenizer = re.compile('\w+')
regex_tokenizer = re.compile('\w+[-.\']?\w+|\S') ## FILL IN THE ANSWER

# look if our dummy example is now tokenized properly
print(dummy_example)
print(re.findall(regex_tokenizer, dummy_example))

It's a dum-dum example, we'll place it here to prove a point. Also, look at this number: 300.99.
["It's", 'a', 'dum-dum', 'example', ',', "we'll", 'place', 'it', 'here', 'to', 'prove', 'a', 'point', '.', 'Also', ',', 'look', 'at', 'this', 'number', ':', '300.99', '.']


In [58]:
# checks if punctuation marks are separated from the end of a word
assert_equal(re.findall(regex_tokenizer, "hello!.?"),['hello', '!', '.', '?'])

# checks if punctuation marks are separated from the end of a word and words are whitespace separated
assert_equal(re.findall(regex_tokenizer, "well,  well, well"),['well', ',', 'well', ',', 'well'])

# checks if words and numbers with -'. inside are kept intact
assert_equal(re.findall(regex_tokenizer, "bye-bye, we'll call you at 3.15"),
             ['bye-bye', ',', "we'll", 'call', 'you', 'at', '3.15'])

# checks if the last token in the dummy sentence is correct
assert_equal(re.findall(regex_tokenizer, dummy_example)[-1], '.')

# checks if number of tokens in the dummy sentence is correct
assert_equal(len(re.findall(regex_tokenizer, dummy_example)), 23)


## 2.3 <a class="anchor" id="subtask_2_3"></a>
### Use Treebank tokenizer (1 points)

As you've already noticed, the process of creating a tokenizer is pretty complicated. There are many more things to consider, and a tokenizer should be chosen in accordance with a task. For example, we might also want to capture abbreviations (U.S.A.), percentages (82%) or URLs.

Luckily, there are already several good tokenizers implemented for us. For instance, the NLTK package has several (https://www.nltk.org/api/nltk.tokenize.html#module-nltk.tokenize). 

Let's tokenize our text using the Treebank tokenizer. It uses regular expressions to tokenize text so that tokens match those used in a popular [Penn Treebank](https://web.archive.org/web/19970614160127/http://www.cis.upenn.edu/~treebank/) dataset. This tokenizer also assumes that the text has already been segmented into sentences. Perform sentence segmentation using NLTK's `sent_tokenize()`. Don't forget to lowercase the tokens after tokenizing. If you lowercase before tokenization, it may have an effect on the segmentation algorithm.

In [59]:
import nltk
nltk.download('punkt')
from nltk.tokenize import sent_tokenize, TreebankWordTokenizer

# here is how the Treebank tokenizer handles the dummy sentence
print(TreebankWordTokenizer().tokenize(dummy_example))

def tokenize_and_lowercase(file_name):
    with open(file_name, 'r') as fr:
        string=fr.read()
    
    """This function tokenizes text files into lowercased tokens with TreebankWordTokenizer
    
    Read a text file into a string,
    tokenize this string into sentences using sent_tokenize(),
    tokenize each sentence into tokens using TreebankWordTokenizer().tokenize(),
    lowercase each token


    Parameters
    ----------
    file_name : str
        a path to the text file
    
    Returns
    -------
    tokens : list of strings
        text as a list of lowercased tokens
    """
    # YOUR CODE HERE
    sentences=sent_tokenize(string)
    tokens=list()
    for sentence in sentences:
        token_list=TreebankWordTokenizer().tokenize(sentence)
        for token in token_list:
            tokens.append(token.lower())
    return tokens


tokenized_bug = tokenize_and_lowercase('/coursedata/01-text-processing/the_gold-bug.txt')

['It', "'s", 'a', 'dum-dum', 'example', ',', 'we', "'ll", 'place', 'it', 'here', 'to', 'prove', 'a', 'point.', 'Also', ',', 'look', 'at', 'this', 'number', ':', '300.99', '.']


[nltk_data] Downloading package punkt to /home/jarvint12/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [60]:
# checks if the number of tokens is correct
assert_equal(len(tokenize_and_lowercase('/coursedata/01-text-processing/the_gold-bug.txt')), 16756)
# checks if all tokens are lowercased
assert_equal(all([x.lower for x in tokenize_and_lowercase('/coursedata/01-text-processing/the_gold-bug.txt')]), True)
# checks if the first token is correct
assert_equal(tokenize_and_lowercase('/coursedata/01-text-processing/the_gold-bug.txt')[0], 'the')
# checks if the last 5 tokens are correct
assert_equal(tokenize_and_lowercase('/coursedata/01-text-processing/the_gold-bug.txt')[-5:], ['who', 'shall', 'tell', '?', '”'])



## TASK 3 <a class="anchor" id="task_3"></a>
## Word frequencies
In this task we will explore the distribution of word frequencies and discuss what influence it can have on different NLP tasks.

## 3.1 <a class="anchor" id="subtask_3_1"></a>
### Analyse word frequencies (3 points)
You've already recorded the statistics of letters and letter pairs, now you can repurpose those functions to answer the following questions:

1. How many word tokens are there in the text? (0.6 points)
2. How many word types are there in the text? (0.6 points)
3. What are 10 most frequent words? Report them as a list starting with the most frequent one? (0.6 points)
4. What is the fraction of word types (out of all word types) that appeared in the text only 2 times or less? (0.6 points)
5. What is the fraction of word types (out of all word types) that appeared in the text 50 times or more? (0.6 points)

Type your answers in the cell below. You can create an additional cell to do the calculations if needed.

In [61]:
# YOUR CODE HERE

# put your answer to question 1 as an int into the variable below
# For example:
# n_word_tokens = 1000
n_word_tokens = len(tokenized_bug) ##FILL IN THE ANSWER
print(n_word_tokens)
# put your answer to question 3 as an int into the variable below
# For example:
# n_word_types = 1000
n_word_types = len(set(tokenized_bug)) ##FILL IN THE ANSWER remove duplicates, 
print(n_word_types)

# put your answer to question 3 as list of strings into the variable below
# For example:
# top_ten_words = ['hello', 'world']
word_counts=collections.Counter()
for word in tokenized_bug:
    word_counts[word]+=1
sorted_list=sorted(word_counts.items(), key=lambda pair: pair[1], reverse=True)
top_ten_words=list()
i=0
while(i<10):
    top_ten_words.append(sorted_list[i][0])
    i+=1
    ##FILL IN THE ANSWER
print(top_ten_words)

# put your answer to question 4 as a float into the variable below
# For example:
sum_rare=0
sum_all=0
for pair in sorted_list:
    if pair[1]<=2:
        sum_rare+=1
    sum_all+=1
frac_2_or_less = sum_rare/sum_all ##FILL IN THE ANSWER

# put your answer to question 5 as a float into the variable below
# For example:
sum_popular=0
for pair in sorted_list:
    if pair[1]>=50:
        sum_popular+=1
##FILL IN THE ANSWER
frac_50_or_more = sum_popular/sum_all ##FILL IN THE ANSWER

16756
2845
[',', 'the', '.', 'of', 'and', 'i', 'a', 'to', '“', 'in']


In [62]:
# checks if your answer is an int
assert_equal(type(n_word_tokens),int)

In [63]:
# checks if your answer is an int
assert_equal(type(n_word_types),int)

In [64]:
# checks if your answer is of the right formt
assert_equal(type(top_ten_words),list)
assert_equal(type(top_ten_words[0]),str)
assert_equal(len(top_ten_words),10)
# checks if you've got the most frequent token right
assert_equal(top_ten_words[0],',')


In [65]:
# checks if your answer is a float
assert_equal(type(frac_2_or_less),float)

In [66]:
# checks if your answer is a float
assert_equal(type(frac_50_or_more),float)

## 3.2 <a class="anchor" id="subtask_3_2"></a>
### Remove stop words (1 point)
As you can see, the most frequent wrod types are not specific to the Poe's story, but are pretty much the same across English language.  

In information theory, the more likely an event to occur, the less information it contains. Thus, if an event is not a surprise, it's simply "old news". For some natural language applications, it means that words like *to* and *the* don't tell anything important about a text. They are not helpful in recognising its topic or its author, for instance.

Such frequent uninformative words are called **stop words**, and, in some cases, they can simply be cleaned out from data. There exist prepared lists of such words in English. Let's remove the ones provided by NLTK.

In [67]:
from nltk.corpus import stopwords
nltk.download('stopwords')

stop_words_english = stopwords.words('english')
#print(stop_words_english)

def remove_stop_words(tokenized_text, stop_words):
    """This function removes stop words from lowercased tokenized text
    
    Parameters
    ----------
    tokenized_text : list of strings
        lowercased text tokens 
    stop_words : list of strings
        a list of words to remove
    
    Returns
    -------
    clean_text : list of strings
        list of text tokens with stop words removed
    """
    
    # YOUR CODE HERE
    #print(tokenized_text)
    clean_text=list()
    for token in tokenized_text:
        if not token in stop_words:
            clean_text.append(token)
    
    return clean_text
    
clean_bug = remove_stop_words(tokenized_bug, stop_words_english)

[nltk_data] Downloading package stopwords to
[nltk_data]     /home/jarvint12/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [68]:
# checks if the number of tokens is right
assert_equal(len(clean_bug), 9630)

# checks if the first token is right
assert_equal(clean_bug[0], 'gold-bug')


## 3.3 <a class="anchor" id="subtask_3_3"></a>
### How much is  left without stop words? (1 point)
Great! Now you have a clean text that can further be used for such tasks as sentiment analysis, for example. 

The last thing. What was the fraction of stop word tokens in the tokenized text? Write your answer in the cell below.

In [69]:
# YOUR CODE HERE

# put your answer as a float into the variable below
# For example:
# fraction_of_stop_word_tokens = 0.5
fraction_of_stop_word_tokens = (len(tokenized_bug)-len(clean_bug))/len(tokenized_bug) ## FILL IN THE ANSWER
print(fraction_of_stop_word_tokens)

0.4252804965385534


In [70]:
# checks if your answer is float
assert_equal(type(fraction_of_stop_word_tokens),float)

## Checklist before submission <a class="anchor" id="checklist"></a>
### 1
To make sure that you didn't forget to import some package or to name some variable, press **Kernel -> Restart** and then **Cell -> Run All**. This way your code will be run exactly in the same order as during the autograding.
### 2
Click the **Validate** button in the upper menu to check that you haven't missed anything.
### 3
To submit the notebook, click on the **jupyterhub** logo in the upper left part of the window, choose the **Assignments** folder, and press **submit**. You can submit multiple times, only the last one counts.
### 4
Please fill in the feedback form in the [Assignment](https://mycourses.aalto.fi/mod/questionnaire/view.php?id=689919) section of Mycoures.